packages/vite/src/node/optimizer/rollupDepPlugin.ts

```diff
@@ -0,0 +1,323 @@
+  'jsx',
+  'tsx',
+  ...KNOWN_ASSET_TYPES,
+]
+
+export function rollupDepPlugin(
+  qualified: Record<string, string>,
+  external: string[],
+  config: ResolvedConfig,
+  ssr: boolean,
+): Plugin {
+  const { extensions } = getDepOptimizationConfig(config, ssr)
+
+  // remove optimizable extensions from `externalTypes` list
+  const allExternalTypes = extensions
+    ? externalTypes.filter((type) => !extensions?.includes('.' + type))
+    : externalTypes
+
+  // use separate package cache for optimizer as it caches paths around node_modules
+  // and it's unlikely for the core Vite process to traverse into node_modules again
+  const esmPackageCache: PackageCache = new Map()
+  const cjsPackageCache: PackageCache = new Map()
+
+  // default resolver which prefers ESM
+  const _resolve = config.createResolver({
+    asSrc: false,
+    scan: true,
+    packageCache: esmPackageCache,
+  })
+
+  // cjs resolver that prefers Node
+  const _resolveRequire = config.createResolver({
+    asSrc: false,
+    isRequire: true,
+    scan: true,
+    packageCache: cjsPackageCache,
+  })
+
+  const resolve = (
+    id: string,
+    importer: string | undefined,
+    kind: ImportKind,
+    resolveDir?: string,
+  ): Promise<string | undefined> => {
+    let _importer: string | undefined = undefined
+    // explicit resolveDir - this is passed only during yarn pnp resolve for
+    // entries
+    if (resolveDir) {
+      _importer = normalizePath(path.join(resolveDir, '*'))
+    } else if (importer) {
+      // map importer ids to file paths for correct resolution
+      _importer = importer in qualified ? qualified[importer] : importer
+    }
+    const resolver = kind.startsWith('require') ? _resolveRequire : _resolve
+    return resolver(id, _importer, undefined, ssr)
+  }
+
+  const resolveResult = (id: string, resolved: string) => {
+    if (resolved.startsWith(browserExternalId)) {
+      return {
+        id: browserExternalNamespace + id,
+      }
+    }
+    if (resolved.startsWith(optionalPeerDepId)) {
+      return {
+        id: optionalPeerDepNamespace + resolved,
+      }
+    }
+    if (ssr && isBuiltin(resolved)) {
+      return
+    }
+    if (isExternalUrl(resolved)) {
+      return {
+        id: resolved,
+        external: true,
+      }
+    }
+    return {
+      id: path.resolve(resolved),
+    }
+  }
+
+  const allExternalTypesReg = new RegExp(
+    `\\.(` + allExternalTypes.join('|') + `)(\\?.*)?$`,
+  )
+
+  function resolveEntry(id: string) {
+    const flatId = flattenId(id)
+    if (flatId in qualified) {
+      return {
+        id: qualified[flatId],
+      }
+    }
+  }
+
+  return {
+    name: 'vite:dep-pre-bundle',
+    // clear package cache when build is finished
+    buildEnd() {
+      esmPackageCache.clear()
+      cjsPackageCache.clear()
+    },
+    resolveId: async function (id, importer, options) {
+      // TODO rolldown: kind is not available in rollup
+      const kind: ImportKind = options.custom?.kind
```

Comment:
Rolldown should support `ImportKind`, see https://github.com/rolldown-rs/rolldown/issues/166